log/src/main/java/kcl/teamIndexZero/traffic/log/FileIO/FileInput.java

```diff
@@ -73,11 +73,31 @@ public FileInput(String folder_path, String file_name) throws InvalidPathExcepti
     }
 
     /**
+     * Closes the reader
+     *
+     * @return Success
+     */
+    public synchronized boolean closeReader() {
+        if (this.reader != null) {
+            try {
+                reader.close();
+                this.reader = null;
+                return true;
+            } catch (IOException e) {
+                MicroLogger.INSTANCE.log_Error("IOException raised in [FileInput.closeReader()] for , ", super.getFilePath());
+                MicroLogger.INSTANCE.log_ExceptionMsg(e);
+                return false;
```

Comment:
A bit of unusual style for java: basically we have 1 normal flow (reader closed, all good) and one (may be more) faulty flows - i.e. the device has been disconnected and we're now in an inconsistent state or something. In java, normally instead of what's done here (return codes true/false to indicate the success of operation) you would use exceptions, that's exactly why: 

1. Information about the fault and causal relativity. Usually and normally, exceptions are chained, i.e. in the top code you'd catch this exception, do what you can about it (recovery attempt) and then rethrow another exception, which takes this one as a constructor. As a result, the top exception will have 'this' exception as cause, and it may be tracked back to it. 
   1. Well, you may object it is still written into the log. Problem is, we have it in log too, but we are losing causal relationship. Consider, this log library is used alongside another library which does not work with log. I.e. has its own log. That lib would be only able to write to its log that 'something failed while closing reader', and this MicrLogger message goes to the other file not directly related to the other one. Generally what we try to use here - least common denominator - which is a language feature 'exception' - to pass as much information about faults as possible.
2. Simplicity of handling. Error handling code is decoupled from the actual logic code. I.e. in the calling code you have no need for the branching, actually it also goes in a 'normal' flow until the exception is thrown, where it passes to its own console block. 
   1. Basically approach is like on each level you are doing what you can, and then either considering something 'recovered, all done' -> terminate exception, or think someone on top may also need that -> rethrow wrapped exception. 

Again - I don't think we should change something right now, we're good to go with what we have here. 
We could talk about this more on our Wednesday session.